File: backend/app/schemas/production.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)


# ============================================================
# PRODUCTION ORDER
# ============================================================


class ProductionOrderCreate(BaseModel):
    """
    Manual/internal Production Order creation schema.

    Manufactured output is identified by product_name.
    It does not need to exist in purchased Products/Stock.
    """

    proforma_id: int = Field(
        gt=0,
    )

    proforma_item_id: int | None = Field(
        default=None,
        gt=0,
    )

    product_name: str = Field(
        min_length=1,
        max_length=500,
    )

    unit: str = Field(
        default="Nos",
        min_length=1,
        max_length=50,
    )

    # Legacy purchased-product relationship only.
    product_id: int | None = Field(
        default=None,
        gt=0,
    )

    quantity: int = Field(
        gt=0,
    )

    planned_start_date: date | None = None

    notes: str | None = None


class ProductionOrderUpdate(BaseModel):
    """
    Update Production Order planning fields.

    Manufactured product identity is independent from
    purchased Products/Stock.
    """

    product_name: str | None = Field(
        default=None,
        min_length=1,
        max_length=500,
    )

    unit: str | None = Field(
        default=None,
        min_length=1,
        max_length=50,
    )

    product_id: int | None = Field(
        default=None,
        gt=0,
    )

    quantity: int | None = Field(
        default=None,
        gt=0,
    )

    planned_start_date: date | None = None

    notes: str | None = None


class ProductionOrderResponse(BaseModel):

    id: int

    production_number: str

    proforma_id: int

    proforma_item_id: int | None = None

    product_name: str

    unit: str

    # Legacy only.
    product_id: int | None = None

    quantity: int

    status: str

    planned_start_date: date | None = None
    actual_start_date: date | None = None
    actual_end_date: date | None = None

    notes: str | None = None

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


# ============================================================
# PRODUCTION MATERIAL
# ============================================================


class ProductionMaterialCreate(BaseModel):

    product_id: int | None = None

    material_name: str = Field(
        min_length=1,
        max_length=200,
    )

    unit: str | None = Field(
        default=None,
        max_length=30,
    )

    quantity_required: Decimal = Field(
        gt=0,
        max_digits=12,
        decimal_places=2,
    )

    unit_cost: Decimal = Field(
        default=Decimal("0.00"),
        ge=0,
        max_digits=14,
        decimal_places=2,
    )


class ProductionMaterialUpdate(BaseModel):

    product_id: int | None = None

    material_name: str | None = Field(
        default=None,
        min_length=1,
        max_length=200,
    )

    unit: str | None = Field(
        default=None,
        max_length=30,
    )

    quantity_required: Decimal | None = Field(
        default=None,
        gt=0,
        max_digits=12,
        decimal_places=2,
    )

    unit_cost: Decimal | None = Field(
        default=None,
        ge=0,
        max_digits=14,
        decimal_places=2,
    )


class ProductionMaterialResponse(BaseModel):

    id: int
    production_order_id: int

    # This IS allowed to reference purchased Stock,
    # because Production Materials are materials consumed.
    product_id: int | None = None

    material_name: str

    unit: str | None = None

    quantity_required: Decimal
    quantity_issued: Decimal

    unit_cost: Decimal
    material_cost: Decimal

    model_config = ConfigDict(
        from_attributes=True,
    )


class ProductionMaterialSummaryResponse(BaseModel):

    production_order_id: int

    total_materials: int

    total_quantity_required: Decimal
    total_quantity_issued: Decimal
    total_quantity_remaining: Decimal

    total_material_cost: Decimal


# ============================================================
# PRODUCTION OPERATION
# ============================================================


class ProductionOperationCreate(BaseModel):

    operation_name: str = Field(
        min_length=1,
        max_length=100,
    )

    machine_name: str | None = Field(
        default=None,
        max_length=150,
    )

    hourly_rate: Decimal = Field(
        default=Decimal("0.00"),
        ge=0,
        max_digits=12,
        decimal_places=2,
    )

    planned_hours: Decimal = Field(
        default=Decimal("0.00"),
        ge=0,
        max_digits=10,
        decimal_places=2,
    )


class ProductionOperationUpdate(BaseModel):

    operation_name: str | None = Field(
        default=None,
        min_length=1,
        max_length=100,
    )

    machine_name: str | None = Field(
        default=None,
        max_length=150,
    )

    hourly_rate: Decimal | None = Field(
        default=None,
        ge=0,
        max_digits=12,
        decimal_places=2,
    )

    planned_hours: Decimal | None = Field(
        default=None,
        ge=0,
        max_digits=10,
        decimal_places=2,
    )


class ProductionOperationComplete(BaseModel):

    actual_hours: Decimal = Field(
        gt=0,
        max_digits=10,
        decimal_places=2,
    )


class ProductionOperationResponse(BaseModel):

    id: int
    production_order_id: int

    operation_name: str

    machine_name: str | None = None

    hourly_rate: Decimal
    planned_hours: Decimal
    actual_hours: Decimal

    operation_cost: Decimal

    status: str

    started_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = ConfigDict(
        from_attributes=True,
    )


class ProductionOperationSummaryResponse(BaseModel):

    production_order_id: int

    total_operations: int

    pending_operations: int
    in_progress_operations: int
    completed_operations: int

    total_planned_hours: Decimal
    total_actual_hours: Decimal

    total_operation_cost: Decimal


# ============================================================
# PRODUCTION ORDER DETAIL
# ============================================================


class ProductionOrderDetailResponse(
    ProductionOrderResponse
):

    materials: list[
        ProductionMaterialResponse
    ] = Field(
        default_factory=list,
    )

    operations: list[
        ProductionOperationResponse
    ] = Field(
        default_factory=list,
    )

File: frontend/src/types/production.ts
export interface ProductionOrder {
  id: number;

  production_number: string;

  proforma_id: number;

  proforma_item_id: number | null;

  /*
   * Manufactured output source of truth.
   */
  product_name: string;

  unit: string;

  /*
   * Legacy purchased Product link only.
   * New manufactured jobs will normally be null.
   */
  product_id: number | null;

  quantity: number;

  status: string;

  planned_start_date: string | null;
  actual_start_date: string | null;
  actual_end_date: string | null;

  notes: string | null;

  created_at: string;
  updated_at: string;
}


export interface ProductionMaterial {
  id: number;

  production_order_id: number;

  /*
   * Materials may reference purchased Stock products.
   */
  product_id: number | null;

  material_name: string;

  unit: string | null;

  quantity_required: string;
  quantity_issued: string;

  unit_cost: string;
  material_cost: string;
}


export interface ProductionOperation {
  id: number;

  production_order_id: number;

  operation_name: string;

  machine_name: string | null;

  hourly_rate: string;

  planned_hours: string;
  actual_hours: string;

  operation_cost: string;

  status: string;

  started_at: string | null;
  completed_at: string | null;
}


export interface ProductionOrderDetail
  extends ProductionOrder {

  materials: ProductionMaterial[];

  operations: ProductionOperation[];
}


export interface ProductionMaterialSummary {
  production_order_id: number;

  total_materials: number;

  total_quantity_required: string;
  total_quantity_issued: string;
  total_quantity_remaining: string;

  total_material_cost: string;
}


export interface ProductionOperationSummary {
  production_order_id: number;

  total_operations: number;

  pending_operations: number;
  in_progress_operations: number;
  completed_operations: number;

  total_planned_hours: string;
  total_actual_hours: string;

  total_operation_cost: string;
}


/* =========================================================
   PRODUCTION ORDER PAYLOADS
========================================================= */

export interface ProductionOrderCreatePayload {
  proforma_id: number;

  proforma_item_id: number | null;

  product_name: string;

  unit: string;

  product_id: number | null;

  quantity: number;

  planned_start_date: string | null;

  notes: string | null;
}


export interface ProductionOrderUpdatePayload {
  product_name?: string;

  unit?: string;

  product_id?: number | null;

  quantity?: number;

  planned_start_date?: string | null;

  notes?: string | null;
}


/* =========================================================
   MATERIAL PAYLOADS
========================================================= */

export interface ProductionMaterialCreatePayload {
  product_id: number | null;

  material_name: string;

  unit: string | null;

  quantity_required: number;

  unit_cost: number;
}


export interface ProductionMaterialUpdatePayload {
  product_id?: number | null;

  material_name?: string;

  unit?: string | null;

  quantity_required?: number;

  unit_cost?: number;
}


/* =========================================================
   OPERATION PAYLOADS
========================================================= */

export interface ProductionOperationCreatePayload {
  operation_name: string;

  machine_name: string | null;

  hourly_rate: number;

  planned_hours: number;
}


export interface ProductionOperationUpdatePayload {
  operation_name?: string;

  machine_name?: string | null;

  hourly_rate?: number;

  planned_hours?: number;
}


export interface ProductionOperationCompletePayload {
  actual_hours: number;
}